pkg/submitter/submitter.go

```diff
@@ -0,0 +1,510 @@
+						s.saveGasUsed(ctx, tx)
+						s.submitCount.Inc()
+						// A solution has been submitted so the
+						// pending solution doesn't matter here any more so reset it.
+						s.solutionPending = nil
+						return
+					}
+				}
+
+				// At this point it means that profit was not enough so need to retry in some time.
+				level.Debug(s.logger).Log("msg", "profit was not enough, retry in 60 seconds", "account", s.account.Address.String())
+				retryTicker := time.NewTicker(60 * time.Second)
+				select {
+				case <-ctx.Done(): // for any blocking operation always check the main context for an early exit.
+				case <-retryTicker.C:
+					continue // Retry send.
+				}
+			}
+		}
+	}(ctx, result)
+}
+
+func (s *Submitter) Submit(ctx context.Context, result *mining.Result) (*types.Transaction, error) {
+	challenge := result.Work.Challenge
+	nonce := result.Nonce
+	s.currentChallenge = challenge
+	s.currentNonce = nonce
+
+	// The submit contains values for 5 data IDs so add them here.
+	for i := 0; i < 5; i++ {
+		valKey := fmt.Sprintf("%s%d", db.QueriedValuePrefix, challenge.RequestIDs[i].Uint64())
+		m, err := s.proxy.BatchGet([]string{valKey})
+		if err != nil {
+			return nil, errors.Wrapf(err, "retrieve pricing data for current request id")
+		}
+		val := m[valKey]
+		var value *big.Int
+		if len(val) == 0 {
+			jsonFile, err := os.Open(s.cfg.ManualDataFile)
+			if err != nil {
+				return nil, errors.Wrapf(err, "manualData read Error")
+			}
+			defer jsonFile.Close()
+			byteValue, _ := ioutil.ReadAll(jsonFile)
+			var result map[string]map[string]uint
+			_ = json.Unmarshal([]byte(byteValue), &result)
+			_id := strconv.FormatUint(challenge.RequestIDs[i].Uint64(), 10)
+			val := result[_id]["VALUE"]
+			if val == 0 {
+				return nil, errors.Errorf("retrieve pricing data for current request id")
+			}
+			value = big.NewInt(int64(val))
+		} else {
+			value, err = hexutil.DecodeBig(string(val))
+			if err != nil {
+				if challenge.RequestIDs[i].Uint64() > tracker.MaxPSRID() {
+					level.Error(s.logger).Log(
+						"msg", "decoding price value prior to submitt ing solution",
+						"err", err,
+					)
+					if len(val) == 0 {
+						level.Error(s.logger).Log("msg", "0 value being submitted")
+						s.currentValues[i] = big.NewInt(0)
+					}
+					continue
+				}
+				return nil, errors.Errorf("no value in database,  reg id:%v", challenge.RequestIDs[i].Uint64())
+			}
+		}
+		s.currentValues[i] = value
+	}
+
+	lastSubmit, err := s.lastSubmit()
+	if err != nil {
+		level.Error(s.logger).Log("msg", "checking last submit time", "err", err)
+	} else if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+		level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+		return nil, errors.New("min transaction submit threshold hasn't passed")
+	}
```

Comment:
This looks like a duplicate check since we already check for this in `handleSubmit`
```suggestion
	lastSubmit, err := s.lastSubmit()
	if err != nil {
		level.Error(s.logger).Log("msg", "checking last submit time", "err", err)
	} else if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
		level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
		return nil, errors.New("min transaction submit threshold hasn't passed")
	}
```

Reply:
Thanks. removed!